src/storage/compression/zstd.cpp

```diff
@@ -0,0 +1,1052 @@
+
+struct ZSTDAnalyzeState : public AnalyzeState {
+public:
+	ZSTDAnalyzeState(CompressionInfo &info, DBConfig &config) : AnalyzeState(info), config(config), context(nullptr) {
+		context = duckdb_zstd::ZSTD_createCCtx();
+	}
+	~ZSTDAnalyzeState() override {
+		duckdb_zstd::ZSTD_freeCCtx(context);
+	}
+
+public:
+	inline void AppendString(const string_t &str) {
+		auto string_size = str.GetSize();
+		total_size += string_size;
+	}
+
+public:
+	DBConfig &config;
+
+	duckdb_zstd::ZSTD_CCtx *context;
+	//! The combined string lengths for all values in the segment
+	idx_t total_size = 0;
+	//! The total amount of values in the segment
+	idx_t count = 0;
+
+	//! The amount of vectors per filled segment
+	idx_t vectors_per_segment = 0;
+	//! The total amount of segments we will create
+	idx_t segment_count = 0;
+	//! Current vector in the segment
+	idx_t vectors_in_segment = 0;
+	//! Current amount of values in the vector
+	idx_t values_in_vector = 0;
+};
+
+unique_ptr<AnalyzeState> ZSTDStorage::StringInitAnalyze(ColumnData &col_data, PhysicalType type) {
+	CompressionInfo info(col_data.GetBlockManager().GetBlockSize());
+	auto &data_table_info = col_data.info;
+	auto &attached_db = data_table_info.GetDB();
+	auto &config = DBConfig::Get(attached_db);
+
+	return make_uniq<ZSTDAnalyzeState>(info, config);
+}
+
+bool ZSTDStorage::StringAnalyze(AnalyzeState &state_p, Vector &input, idx_t count) {
+	auto &state = state_p.Cast<ZSTDAnalyzeState>();
+	UnifiedVectorFormat vdata;
+	input.ToUnifiedFormat(count, vdata);
+
+	auto data = UnifiedVectorFormat::GetData<string_t>(vdata);
+	for (idx_t i = 0; i < count; i++) {
+		auto idx = vdata.sel->get_index(i);
+		if (!vdata.validity.RowIsValid(idx)) {
+			continue;
+		}
+		auto &str = data[idx];
+		auto string_size = str.GetSize();
+		state.total_size += string_size;
+	}
+	state.values_in_vector += count;
+	while (state.values_in_vector >= ZSTD_VECTOR_SIZE) {
+		if (GetVectorMetadataSize(state.vectors_in_segment + 1) > GetWritableSpace(state.info)) {
+			state.vectors_per_segment = state.vectors_in_segment;
+			state.segment_count++;
+			state.vectors_in_segment = 0;
+		} else {
+			state.vectors_in_segment++;
+		}
+		state.values_in_vector -= ZSTD_VECTOR_SIZE;
+	}
+	state.count += count;
+	return true;
+}
+
+// Compression score to determine which compression to use
+idx_t ZSTDStorage::StringFinalAnalyze(AnalyzeState &state_p) {
+	auto &state = state_p.Cast<ZSTDAnalyzeState>();
+
+	if (!state.count) {
+		return NumericLimits<idx_t>::Maximum();
+	}
+
+	if (state.values_in_vector) {
+		D_ASSERT(state.values_in_vector < ZSTD_VECTOR_SIZE);
+		state.segment_count++;
+	}
+
+	double penalty;
+	idx_t average_length = state.total_size / state.count;
+	auto overflow_string_threshold = StringUncompressed::GetStringBlockLimit(state.info.GetBlockSize());
+	if (average_length < 100) {
+		penalty = 2.0;
+	} else if (average_length >= overflow_string_threshold) {
```

Comment:
* Remove the usage of zstd for lengths < 100
* Configure the `overflow_string_threshold ` - e.g. `SET zstd_min_string_length = 4096`